File: packages/types/src/events/tool-events.ts
/**
 * Tool Related Event Type Definitions
 */

import type { ID } from "../common.js";
import type { BaseEvent } from "./base.js";

/**
 * Tool call start event type
 */
export interface ToolCallStartedEvent extends BaseEvent {
  type: "TOOL_CALL_STARTED";
  /** Node ID */
  nodeId: ID;
  /** Tool ID */
  toolId: ID;
  /** Tool call task ID (for tracking individual tool calls) */
  taskId?: string;
  /** Batch ID (identification of a batch of parallel tool calls) */
  batchId?: string;
  /** Tool name */
  toolName?: string;
  /** Tool parameters */
  toolArguments: string;
}

/**
 * Tool call completion event type
 */
export interface ToolCallCompletedEvent extends BaseEvent {
  type: "TOOL_CALL_COMPLETED";
  /** Node ID */
  nodeId: ID;
  /** Tool ID */
  toolId: ID;
  /** Tool call task ID (for tracking individual tool calls) */
  taskId?: string;
  /** Batch ID (identification of a batch of parallel tool calls) */
  batchId?: string;
  /** Tool name */
  toolName?: string;
  /** Tool results */
  toolResult: unknown;
  /** execution time */
  executionTime: number;
}

/**
 * Tool Call Failure Event Type
 */
export interface ToolCallFailedEvent extends BaseEvent {
  type: "TOOL_CALL_FAILED";
  /** Node ID */
  nodeId: ID;
  /** Tool ID */
  toolId: ID;
  /** Tool call task ID (for tracking individual tool calls) */
  taskId?: string;
  /** Batch ID (identification of a batch of parallel tool calls) */
  batchId?: string;
  /** Tool name */
  toolName?: string;
  /** error message */
  error: string;
}

/**
 * Tool Call Blocked Event Type (NEW - for failure protection)
 */
export interface ToolCallBlockedEvent extends BaseEvent {
  type: "TOOL_CALL_BLOCKED";
  /** Execution ID */
  executionId: ID;
  /** Node ID */
  nodeId: ID;
  /** Tool ID */
  toolId: ID;
  /** Tool name */
  toolName?: string;
  /** Current failure count */
  failureCount: number;
  /** Last error message */
  lastError?: string;
  /** Remaining cooldown time in milliseconds */
  remainingCooldown?: number;
  /** Blocking reason */
  reason?: string;
}

/**
 * Tool to add event types
 */
export interface ToolAddedEvent extends BaseEvent {
  type: "TOOL_ADDED";
  /** Node ID */
  nodeId: ID;
  /** Tool ID List */
  toolIds: ID[];
  /** Scope (EXECUTION or LOCAL) */
  scope: "EXECUTION" | "LOCAL";
  /** Number of tools successfully added */
  addedCount: number;
  /** Number of tools skipped */
  skippedCount: number;
}

/**
 * Tool visibility changed event type
 * Triggered when the set of visible tools changes (scope switch, dynamic add/remove, etc.)
 */
export interface ToolVisibilityChangedEvent extends BaseEvent {
  type: "TOOL_VISIBILITY_CHANGED";
  /** Execution ID */
  executionId: ID;
  /** Workflow ID */
  workflowId?: ID;
  /** Node ID (optional, if triggered by node execution) */
  nodeId?: ID;
  /** Current scope */
  scope: "EXECUTION" | "SUBGRAPH" | "LOOP";
  /** Scope ID */
  scopeId: ID;
  /** Change type */
  changeType: "enter_scope" | "exit_scope" | "add_tools" | "remove_tools" | "refresh" | "init";
  /** Currently visible tool IDs */
  visibleToolIds: ID[];
  /** Previous visible tool IDs (for comparison) */
  previousVisibleToolIds?: ID[];
  /** Timestamp of the change */
  timestamp: number;
}

File: crates/wf-types/src/events/tool_events.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolCallStartedEvent {
    pub base: super::BaseEvent,
    pub node_id: super::super::Id,
    pub tool_id: super::super::Id,
    pub task_id: Option<String>,
    pub batch_id: Option<String>,
    pub tool_name: Option<String>,
    pub tool_arguments: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolCallCompletedEvent {
    pub base: super::BaseEvent,
    pub node_id: super::super::Id,
    pub tool_id: super::super::Id,
    pub task_id: Option<String>,
    pub batch_id: Option<String>,
    pub tool_name: Option<String>,
    pub tool_result: serde_json::Value,
    pub execution_time: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolCallFailedEvent {
    pub base: super::BaseEvent,
    pub node_id: super::super::Id,
    pub tool_id: super::super::Id,
    pub task_id: Option<String>,
    pub batch_id: Option<String>,
    pub tool_name: Option<String>,
    pub error: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolCallBlockedEvent {
    pub base: super::BaseEvent,
    pub execution_id: super::super::Id,
    pub node_id: super::super::Id,
    pub tool_id: super::super::Id,
    pub tool_name: Option<String>,
    pub failure_count: u32,
    pub last_error: Option<String>,
    pub remaining_cooldown: Option<i64>,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolAddedEvent {
    pub base: super::BaseEvent,
    pub node_id: super::super::Id,
    pub tool_ids: Vec<super::super::Id>,
    pub scope: String,
    pub added_count: u32,
    pub skipped_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolVisibilityChangedEvent {
    pub base: super::BaseEvent,
    pub execution_id: super::super::Id,
    pub workflow_id: Option<super::super::Id>,
    pub node_id: Option<super::super::Id>,
    pub scope: String,
    pub scope_id: super::super::Id,
    pub change_type: String,
    pub visible_tool_ids: Vec<super::super::Id>,
    pub previous_visible_tool_ids: Option<Vec<super::super::Id>>,
    pub timestamp: i64,
}
